aten/src/ATen/core/TensorImpl.h

```diff
@@ -724,6 +724,17 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
     storage_offset_ = storage_offset;
   }
 
+  /* Sets the storage of this tensor to be new_storage */
+  void set_storage(const Storage& new_storage) {
+    auto* new_storage_ = new_storage.unsafeGetStorageImpl();
+    auto* old_storage_ = storage_.unsafeGetStorageImpl();
+    AT_ASSERTM(old_storage_, "Tensor: invalid null storage");
+    if (new_storage_ == old_storage_) {
+      return;
+    }
+    storage_ = new_storage;
```

Comment:
Huh. I wonder if `intrusive_ptr` assignment does the same-identity optimization.

Reply:
Nope we don't. @smessmer, this seems worthwhile, right? Saves you a refcount twiddle.

Reply:
I'm not sure it's worth it. Self assignment likely happens rarely and adding a branch has a cost for all calls, not just for the self-assignment calls. The branch predictor (at least on x86) is probably good enough to optimize for the non-self-assignment case, but then once a self-assignment happens, the cost of the branch misprediction might well be larger than the cost of atomically incrementing and decrementing the refcounter.